node/src/components/chain_synchronizer/operations.rs

```diff
@@ -146,17 +146,29 @@ where
             trie_fetch_limit: Semaphore::new(config.max_parallel_trie_fetches()),
         };
 
-        let trusted_block_header = match config.trusted_hash() {
-            Some(trusted_hash) => {
-                *fetch_and_store_initial_trusted_block_header(&ctx, metrics, trusted_hash).await?
+        // The config may contain the hash of a block that is known to be on the correct chain. We
+        // Also assume that all blocks in storage are correct. Fast sync will use whichever is more
+        // recent as a trusted starting point.
+        let maybe_stored_header = effect_builder.get_highest_block_header_from_storage().await;
+        let maybe_config_header = if let Some(trusted_hash) = config.trusted_hash() {
+            Some(*fetch_and_store_initial_trusted_block_header(&ctx, metrics, trusted_hash).await?)
+        } else {
+            None
+        };
+        let trusted_block_header = match (maybe_config_header, maybe_stored_header) {
+            (Some(config_header), None) => config_header,
+            (None, Some(stored_header)) => stored_header,
+            (None, None) => {
+                debug!("no highest block header found in storage, no trusted header configured");
+                return Ok(None);
             }
-            None => match effect_builder.get_highest_block_header_from_storage().await {
-                Some(block_header) => block_header,
-                None => {
-                    debug!("no highest block header found in storage");
-                    return Ok(None);
+            (Some(config_header), Some(highest_header)) => {
+                if config_header.height() > highest_header.height() {
```

Comment:
I'd make it `>=`, so that the trusted hash takes priority if both are at the same height - I think that would be more correct. Not that important, though.

Reply:
Hmm… if they're at the same height but different, I think the right thing to do would be to return an error. Maybe we can do that in another PR?

Reply:
That's also an option, and probably the best one right now, actually. But if we ever decide to support correcting being on a wrong fork of the chain, it might be easier to just prioritize the trusted hash.

I'm fine with doing it in another PR :+1: I could even do it in #3280 (I'm going to merge yours into it once it's in `dev`).